jgrapht-core/src/main/java/org/jgrapht/alg/vertexcover/TreeVCImpl.java

```diff
@@ -0,0 +1,193 @@
+ * started once the method gets invoked.
+ *
+ * @param <V> the graph vertex type.
+ * @param <E> the graph edge type.
+ *
+ * @author Alexandru Valeanu
+ */
+public class TreeVCImpl<V, E> implements VertexCoverAlgorithm<V> {
+
+    // Input graph
+    private final Graph<V, E> graph;
+
+    // Input set of roots (one root per tree)
+    private final Set<V> roots;
+
+    // Cached minimum vertex cover
+    private VertexCover<V> minVertexCover;
+
+    /**
+     * Creates a new TreeVCImpl instance.
+     *
+     * <p>
+     *      Note: The constructor does NOT check if the input graph is a valid tree.
+     * </p>
+     *
+     * <p>
+     *      Note: The vertex cover is computed lazily, when needed.
+     * </p>
+     *
+     * @param graph the input graph
+     * @param root the input root
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws NullPointerException if {@code root} is {@code null}
+     * @throws IllegalArgumentException if {@code root} is an invalid vertex
+     */
+    public TreeVCImpl(Graph<V, E> graph, V root){
+        this(graph, Collections.singleton(Objects.requireNonNull(root, "root cannot be null")));
+    }
+
+    /**
+     * Creates a new TreeVCImpl instance.
+     *
+     * <p>
+     *      Note: The constructor does NOT check if the input graph is a valid tree.
+     * </p>
+     *
+     * <p>
+     *     Note: If two roots correspond to the same tree, an error will be thrown.
+     * </p>
+     *
+     * <p>
+     *      Note: The vertex cover is computed lazily, when needed.
+     * </p>
+     *
+     * @param graph the input graph
+     * @param roots the input root
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws NullPointerException if {@code roots} is {@code null}
+     * @throws IllegalArgumentException if {@code roots} is empty
+     * @throws IllegalArgumentException if {@code roots} contains an invalid vertex
+     */
+    public TreeVCImpl(Graph<V, E> graph, Set<V> roots) {
+        this.graph = Objects.requireNonNull(graph, "graph cannot be null");
+        this.roots = Objects.requireNonNull(roots, "set of roots cannot be null");
+
+        if (this.roots.isEmpty())
+            throw new IllegalArgumentException("set of roots cannot be empty");
+
+        if (!this.graph.vertexSet().containsAll(roots))
+            throw new IllegalArgumentException("invalid set of roots");
+    }
+
+    // Populate parent map by iterating through the BFS tree of root
+    private void bfs(V root, Set<V> visited, Map<V, V> parent){
+        BreadthFirstIterator<V, E> bfs = new BreadthFirstIterator<>(graph, root);
+
+        while (bfs.hasNext()) {
+            V u = bfs.next();
+            visited.add(u);
+            V p = bfs.getParent(u);
+
+            if (p != null)
+                parent.put(u, p);
+        }
+    }
+
+    private void computeMinimumVertexCover(){
+        Map<V, V> parent =  new HashMap<>();
```

Comment:
consider renaming to parentMap. (I got a bit confused between parent and parents)